File: app/services/action_focus.py
"""当下关注点：根据已计算的结构产出可展示语境（非交易指令）。"""

from __future__ import annotations

from typing import Literal, Optional

from app.core.models import (
    ActionFocus,
    ActionFocusActiveBi,
    ActionFocusPivotRef,
    ActionFocusPivotSlot,
    ActionFocusRecentDivergence,
    ActionFocusRecentSignal,
    Divergence,
    Pivot,
    Signal,
    Stroke,
)


def _recent_window_size(kline_count: int) -> int:
    return max(18, min(80, int(kline_count * 0.06)))


def _pick_reference_pivot(pivots: list[Pivot], last_idx: int) -> Optional[Pivot]:
    if not pivots:
        return None
    candidates = [p for p in pivots if last_idx >= p.start_idx and last_idx <= p.end_idx]
    if not candidates:
        return max(pivots, key=lambda p: p.end_idx)
    return max(candidates, key=lambda p: p.end_idx - p.start_idx)


def _pivot_to_ref(p: Pivot) -> ActionFocusPivotRef:
    return ActionFocusPivotRef(
        level=p.level,
        zd=p.zd,
        zg=p.zg,
        start_idx=p.start_idx,
        end_idx=p.end_idx,
    )


def _price_vs_pivot(price: float, pivot: Pivot) -> Literal["inside", "above", "below"]:
    if pivot.zd <= price <= pivot.zg:
        return "inside"
    if price > pivot.zg:
        return "above"
    return "below"


def _slot_for_pivots(pivots: list[Pivot], last_idx: int, price: float) -> ActionFocusPivotSlot:
    ref_p = _pick_reference_pivot(pivots, last_idx)
    if ref_p is None:
        return ActionFocusPivotSlot(relation="none", pivot=None)
    return ActionFocusPivotSlot(
        relation=_price_vs_pivot(price, ref_p),
        pivot=_pivot_to_ref(ref_p),
    )


def _latest_divergence_since(divergences: list[Divergence], since_idx: int) -> Optional[Divergence]:
    windowed = [d for d in divergences if d.idx >= since_idx]
    if not windowed:
        return None
    return max(windowed, key=lambda d: d.idx)


def _latest_signal_since(buy: list[Signal], sell: list[Signal], since_idx: int) -> Optional[Signal]:
    merged = [s for s in buy + sell if s.idx >= since_idx]
    if not merged:
        return None
    return max(merged, key=lambda s: s.idx)


def build_action_focus(
    *,
    current_price: float,
    last_bar_index: int,
    kline_count: int,
    zhongshus: list[Pivot],
    zhongshus_lv2: list[Pivot],
    active_bi: Optional[Stroke],
    divergences: list[Divergence],
    buy_signals: list[Signal],
    sell_signals: list[Signal],
) -> ActionFocus:
    recent_bars = _recent_window_size(kline_count) if kline_count else 1
    recent_since = last_bar_index - recent_bars

    active_bi_out: Optional[ActionFocusActiveBi] = None
    if active_bi is not None:
        active_bi_out = ActionFocusActiveBi(
            direction=active_bi.direction,
            start_price=active_bi.start_price,
            end_price=active_bi.end_price,
        )

    d_latest = _latest_divergence_since(divergences, recent_since)
    recent_div: Optional[ActionFocusRecentDivergence] = None
    if d_latest is not None:
        recent_div = ActionFocusRecentDivergence(
            level=d_latest.level,
            direction=d_latest.direction,
            idx=d_latest.idx,
            ratio=d_latest.ratio,
            structure_kind=d_latest.structure_kind,
        )
    elif divergences:
        d_fallback = max(divergences, key=lambda d: d.idx)
        recent_div = ActionFocusRecentDivergence(
            level=d_fallback.level,
            direction=d_fallback.direction,
            idx=d_fallback.idx,
            ratio=d_fallback.ratio,
            structure_kind=d_fallback.structure_kind,
        )
        recent_bars = max(recent_bars, last_bar_index - d_fallback.idx + 1)

    s_latest = _latest_signal_since(buy_signals, sell_signals, recent_since)
    recent_sig: Optional[ActionFocusRecentSignal] = None
    if s_latest is not None:
        recent_sig = ActionFocusRecentSignal(
            side=s_latest.side,
            kind=s_latest.kind,
            idx=s_latest.idx,
            time=s_latest.time,
            price=s_latest.price,
            stop_loss=s_latest.stop_loss,
            stop_loss_2=s_latest.stop_loss_2,
            take_profit_1=s_latest.take_profit_1,
            take_profit_2=s_latest.take_profit,
        )
    else:
        merged = buy_signals + sell_signals
        if merged:
            s_fallback = max(merged, key=lambda s: s.idx)
            recent_sig = ActionFocusRecentSignal(
                side=s_fallback.side,
                kind=s_fallback.kind,
                idx=s_fallback.idx,
                time=s_fallback.time,
                price=s_fallback.price,
                stop_loss=s_fallback.stop_loss,
                stop_loss_2=s_fallback.stop_loss_2,
                take_profit_1=s_fallback.take_profit_1,
                take_profit_2=s_fallback.take_profit,
            )
            recent_bars = max(recent_bars, last_bar_index - s_fallback.idx + 1)

    return ActionFocus(
        last_bar_index=last_bar_index,
        recent_window_bars=recent_bars,
        current_price=current_price,
        primary_pivot=_slot_for_pivots(zhongshus, last_bar_index, current_price),
        higher_pivot=_slot_for_pivots(zhongshus_lv2, last_bar_index, current_price),
        active_bi=active_bi_out,
        recent_divergence=recent_div,
        recent_signal=recent_sig,
    )

File: frontend/src/action-focus.js
import { KIND_NAME } from "./constants.js";
import { escHtml, structureKindLabel } from "./utils.js";

function nearestPivotContainingLastBar(pivots, lastIdx) {
  if (!pivots.length) return null;
  const candidates = pivots.filter(p => lastIdx >= p.start_idx && lastIdx <= p.end_idx);
  if (!candidates.length) {
    return pivots.reduce((best, p) => (p.end_idx > best.end_idx ? p : best), pivots[0]);
  }
  return candidates.reduce((best, p) => (p.end_idx - p.start_idx > best.end_idx - best.start_idx ? p : best), candidates[0]);
}

function latestDivergenceNearEnd(divergences, sinceIdx) {
  const windowed = divergences.filter(d => d.idx >= sinceIdx);
  if (!windowed.length) return null;
  return windowed.reduce((a, b) => (a.idx > b.idx ? a : b));
}

function latestSignalNearEnd(signals, sinceIdx) {
  const windowed = signals.filter(s => s.idx >= sinceIdx);
  if (!windowed.length) return null;
  return windowed.reduce((a, b) => (a.idx > b.idx ? a : b));
}

function actionFocusHtmlFromApi(actionFocus, activeBiFallback) {
  const px = Number(actionFocus.current_price ?? 0);
  const lastIdx = Number(actionFocus.last_bar_index ?? 0);
  const recentBars = Number(actionFocus.recent_window_bars ?? 18);
  const dirLabel = d => d === "UP" ? "向上" : "向下";

  const slotLine = (label, slot) => {
    if (!slot || slot.relation === "none" || !slot.pivot) return `<div style="margin-top:8px;">${label}：当前无可用中枢参考。</div>`;
    const { zd, zg } = slot.pivot;
    if (slot.relation === "inside") {
      return `<div style="margin-top:8px;">${label}：价格 <b>${px.toFixed(2)}</b> 落在中枢 <b>[${Number(zd).toFixed(2)}, ${Number(zg).toFixed(2)}]</b> 内，属于中枢震荡语境。</div>`;
    }
    if (slot.relation === "above") {
      return `<div style="margin-top:8px;">${label}：价格 <b>${px.toFixed(2)}</b> 在中枢 <b>[${Number(zd).toFixed(2)}, ${Number(zg).toFixed(2)}]</b> 上方，属于偏强离开/回踩观察语境。</div>`;
    }
    return `<div style="margin-top:8px;">${label}：价格 <b>${px.toFixed(2)}</b> 在中枢 <b>[${Number(zd).toFixed(2)}, ${Number(zg).toFixed(2)}]</b> 下方，属于偏弱离开/反抽观察语境。</div>`;
  };

  const active = actionFocus.active_bi || activeBiFallback;
  let activeLine = `<div>未完成笔：暂无（等待新笔启动或结构太短）</div>`;
  if (active) {
    activeLine = `<div>未完成笔：<b>${dirLabel(active.direction)}</b>，${Number(active.start_price).toFixed(2)} → ${Number(active.end_price).toFixed(2)}（仅表示当下这一段<strong>尚未被反向分型确认</strong>）</div>`;
  }

  let divLine = `<div style="margin-top:8px;">最近窗口背驰：最近 ${recentBars} 根内<strong>未发现</strong>可对照的背驰候选。</div>`;
  if (actionFocus.recent_divergence) {
    const d = actionFocus.recent_divergence;
    const lvl = d.level === "bi" ? "笔" : "线段";
    const sk = structureKindLabel(d.structure_kind || "zpan_like");
    divLine = `<div style="margin-top:8px;">最近窗口背驰：<b>${lvl}</b>方向 <b>${d.direction === "DOWN" ? "底背驰语境" : "顶背驰语境"}</b>，<b>${sk}</b>，MACD 比 <b>${Number(d.ratio).toFixed(2)}</b>（仅说明力度对比，不代表此刻应下单）</div>`;
  }

  let hintLine = `<div style="margin-top:8px;color:rgba(232,236,246,.62);">无指令式买卖点，仅为结构语境披露。</div>`;
  if (actionFocus.recent_signal) {
    const s = actionFocus.recent_signal;
    const sideCn = s.side === "BUY" ? "买" : "卖";
    hintLine = `<div style="margin-top:8px;color:#ffbf69;"><b>结构锚点</b>：最近窗口曾标注<b>${KIND_NAME[s.kind] || s.kind}${sideCn}</b>（${s.time}），用于对照<strong>后续</strong>边界测试；<b>不是</b>「此刻必须进场」的指令。</div>`;
  }

  return `
    <div>最新价：<b>${px.toFixed(2)}</b>（最后一根 K 线索引 ${lastIdx}）</div>
    ${activeLine}
    ${slotLine("本级中枢", actionFocus.primary_pivot)}
    ${slotLine("中级别中枢", actionFocus.higher_pivot)}
    ${divLine}
    ${hintLine}
    <div style="margin-top:10px;font-size:11px;color:rgba(232,236,246,.48);">级别选择与风控责任说明见<strong>页脚风险提示</strong>。</div>
  `;
}

export function renderActionFocus(result) {
  const el = document.getElementById("actionFocus");
  if (!result.kline_data || !result.kline_data.length) {
    el.innerHTML = `<div class="muted">暂无 K 线数据</div>`;
    return;
  }
  if (result.action_focus) {
    el.innerHTML = actionFocusHtmlFromApi(result.action_focus, result.active_bi);
    return;
  }
  const lastIdx = result.kline_data.length - 1;
  const px = result.current_price;
  const dirLabel = d => d === "UP" ? "向上" : "向下";

  const recentBars = Math.max(18, Math.min(80, Math.floor(result.kline_data.length * 0.06)));
  const recentSince = lastIdx - recentBars;

  const lastPivot = nearestPivotContainingLastBar(result.zhongshus, lastIdx);
  let posText = "当前价格附近未落到已识别中枢区间内（或中枢列表为空）。";
  if (lastPivot) {
    const { zd, zg } = lastPivot;
    if (px >= zd && px <= zg) {
      posText = `价格 <b>${px.toFixed(2)}</b> 落在最近可视中枢 <b>[${zd.toFixed(2)}, ${zg.toFixed(2)}]</b> 内，属于中枢震荡语境。`;
    } else if (px > zg) {
      posText = `价格 <b>${px.toFixed(2)}</b> 在最近可视中枢 <b>[${zd.toFixed(2)}, ${zg.toFixed(2)}]</b> 上方，属于偏强离开/回踩观察语境。`;
    } else {
      posText = `价格 <b>${px.toFixed(2)}</b> 在最近可视中枢 <b>[${zd.toFixed(2)}, ${zg.toFixed(2)}]</b> 下方，属于偏弱离开/反抽观察语境。`;
    }
  }

  let activeLine = `<div>未完成笔：暂无（等待新笔启动或结构太短）</div>`;
  if (result.active_bi) {
    const ab = result.active_bi;
    activeLine = `<div>未完成笔：<b>${dirLabel(ab.direction)}</b>，${Number(ab.start_price).toFixed(2)} → ${Number(ab.end_price).toFixed(2)}（仅表示当下这一段<strong>尚未被反向分型确认</strong>）</div>`;
  }

  const lastDiv = latestDivergenceNearEnd(result.divergences, recentSince);
  let divLine = `<div style="margin-top:8px;">最近窗口背驰：最近 ${recentBars} 根内<strong>未发现</strong>可对照的背驰候选。</div>`;
  if (lastDiv) {
    const lvl = lastDiv.level === "bi" ? "笔" : "线段";
    const sk = structureKindLabel(lastDiv.structure_kind || "zpan_like");
    divLine = `<div style="margin-top:8px;">最近窗口背驰：<b>${lvl}</b>方向 <b>${lastDiv.direction === "DOWN" ? "底背驰语境" : "顶背驰语境"}</b>，<b>${sk}</b>，MACD 比 <b>${Number(lastDiv.ratio).toFixed(2)}</b>（仅说明力度对比，不代表此刻应下单）</div>`;
  }

  const recentSignal = latestSignalNearEnd([...result.buy_signals, ...result.sell_signals], recentSince);
  let hintLine = `<div style="margin-top:8px;color:rgba(232,236,246,.62);">无指令式买卖点，仅为结构语境披露。</div>`;
  if (recentSignal) {
    const sideCn = recentSignal.side === "BUY" ? "买" : "卖";
    hintLine = `<div style="margin-top:8px;color:#ffbf69;"><b>结构锚点</b>：最近窗口曾标注<b>${KIND_NAME[recentSignal.kind] || recentSignal.kind}${sideCn}</b>（${recentSignal.time}），用于对照<strong>后续</strong>边界测试；<b>不是</b>「此刻必须进场」的指令。</div>`;
  }

  el.innerHTML = `
    <div>最新价：<b>${px.toFixed(2)}</b>（最后一根 K 线索引 ${lastIdx}）</div>
    ${activeLine}
    <div style="margin-top:8px;">${posText}</div>
    ${divLine}
    ${hintLine}
    <div style="margin-top:10px;font-size:11px;color:rgba(232,236,246,.48);">级别选择与风控责任说明见<strong>页脚风险提示</strong>。</div>
  `;
}
